### PR TITLE
Add support for mxfp8 x mxfp4 cutlass MoE for gpt-oss

### PR DESCRIPTION
**Pull request by nvcastet**
## Motivation

`7.2%` e2e perf gain

`18,079.65 tok/s/gpu`

Client cmd:
```
python3 -m sglang.bench_serving --backend sglang --dataset-name random --num-prompt 12800 --random-input 1024 --random-output 2048 --random-range-ratio 1 --max-concurrency 2560
```

With cutlass MoE:
```
python3 -m sglang.launch_server --model openai/gpt-oss-120b --disable-radix-cache  --attention-backend trtllm_mha --tp-size 4 --ep 4 --cuda-graph-max-bs 640 --enable-dp-attention --dp 4 --stream-interval 10 --mem-fraction-static 0.9 --cuda-graph-bs 1 2 4 8 16 24 32 40 48 56 64 72 80 88 96 104 112 120 128 256 512 640 --max-running-requests 2560 --enable-dp-lm-head --enable-flashinfer-cutlass-moe
```

```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 2560
Successful requests:                     12800
Benchmark duration (s):                  362.48
Total input tokens:                      13107200
Total generated tokens:                  26214400
Total generated tokens (retokenized):    25138608
Request throughput (req/s):              35.31
Input token throughput (tok/s):          36159.30
Output token throughput (tok/s):         72318.60
Total token throughput (tok/s):          108477.90
Concurrency:                             2543.76
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   72036.95
Median E2E Latency (ms):                 72074.09
---------------Time to First Token----------------
Mean TTFT (ms):                          6377.65
Median TTFT (ms):                        5959.61
P99 TTFT (ms):                           12209.08
---------------Inter-Token Latency----------------
Mean ITL (ms):                           32.08
Median ITL (ms):                         28.86
P95 ITL (ms):                            58.78
P99 ITL (ms):                            72.18
Max ITL (ms):                            1178.21
==================================================
```

with trtllm MoE
```
python3 -m sglang.launch_server --model openai/gpt-oss-120b --disable-radix-cache  --attention-backend trtllm_mha --tp-size 4 --ep 4 --cuda-graph-max-bs 640 --enable-dp-attention --dp 4 --stream-interval 10 --mem-fraction-static 0.9 --cuda-graph-bs 1 2 4 8 16 24 32 40 48 56 64 72 80 88 96 104 112 120 128 256 512 640 --max-running-requests 2560 --enable-dp-lm-head
```

```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 2560
Successful requests:                     12800
Benchmark duration (s):                  388.46
Total input tokens:                      13107200
Total generated tokens:                  26214400
Total generated tokens (retokenized):    25129042
Request throughput (req/s):              32.95
Input token throughput (tok/s):          33741.82
Output token throughput (tok/s):         67483.65
Total token throughput (tok/s):          101225.47
Concurrency:                             2544.27
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   77213.66
Median E2E Latency (ms):                 77236.78
---------------Time to First Token----------------
Mean TTFT (ms):                          7548.49
Median TTFT (ms):                        7302.39
P99 TTFT (ms):                           14942.58
---------------Inter-Token Latency----------------
Mean ITL (ms):                           34.04
Median ITL (ms):                         30.10
P95 ITL (ms):                            59.72
P99 ITL (ms):                            74.17
Max ITL (ms):                            1441.70
==================================================
```

## Accuracy Tests

```
# python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1319 --parallel 1319
Accuracy: 0.824
Invalid: 0.037
Latency: 17.384 s
Output throughput: 25777.816 token/s
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/layers/quantization/mxfp4.py [hunk 1]
```diff
@@ -644,12 +730,50 @@ def apply(
                         value=0.0,
                     )
             elif self.flashinfer_mxfp4_moe_precision == "default":
-                x_quant, x_scale = mxfp8_quantize(x, False, alignment=self.hidden_size)
+                x_quant, x_scale = mxfp8_quantize(
+                    x,
+                    self.enable_flashinfer_cutlass_moe,
+                    alignment=self.weight_alignment,
+                )
                 x_scale = x_scale.view(torch.float8_e4m3fn).reshape(-1)
```
[the remark this answers is not in the document]
Comment (nvcastet) [reply python/sglang/srt/layers/quantization/mxfp4.py]: There is no `return_e8m0_scale` arg here.

**Comment by Alcanderian** [comment]
Could we add some test？We have b200-8gpu-runner in ci.

**Comment by nvcastet** [comment]
> Could we add some test？We have b200-8gpu-runner in ci.

@Alcanderian Thanks! I just hooked `test_gpt_oss_4gpu.py` to b200 CI and adding extra tests for MoE backends.

**Comment by RonanKMcGovern** [comment]
Thanks for working on this.

Will there be a test for data-parallel compatibility? that would be great

**Comment by nvcastet** [comment]
> Thanks for working on this.
> 
> Will there be a test for data-parallel compatibility? that would be great

Yes the tests added use DP4 + EP4.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 2]
```diff
@@ -730,12 +725,13 @@ def weight_loader_fused(
...
-                param.data.copy_(loaded_weight)
             else:
                 dim1 = loaded_weight.shape[1]
-                dim2 = loaded_weight.shape[2]
-                param.data[:, :dim1, :dim2].copy_(loaded_weight)
+                packing_factor = (
+                    param.data.element_size() // loaded_weight.element_size()
+                )
+                dim2 = loaded_weight.shape[2] // packing_factor
+                param.data[:, :dim1, :dim2].copy_(loaded_weight.view(param.data.dtype))
```
[the remark this answers is not in the document]
Comment (BBuf) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Agree with it, can we refactor two duplicate code into a member function? @nvcastet

[the remark this answers is not in the document]
Comment (nvcastet) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Done

**Comment by Fridge003** [comment]
@nvcastet Can you please rebase this PR and solve the conflicts so we will review it.

**Comment by nvcastet** [comment]
@Fridge003 

> @nvcastet Can you please rebase this PR and solve the conflicts so we will review it.

Done.

File: python/sglang/srt/server_args.py [hunk 3]
```diff
@@ -936,7 +935,9 @@ def _handle_model_specific_adjustments(self):
             )
 
             if is_blackwell_supported() and is_mxfp4_quant_format:
-                self.moe_runner_backend = "flashinfer_mxfp4"
+                if self.moe_runner_backend == "auto":
+                    self.moe_runner_backend = "flashinfer_trtllm"
+                self.quantization = "mxfp4"
                 logger.warning(
                     "Detected SM100 and MXFP4 quantization format for GPT-OSS model, enabling FlashInfer MXFP4 MOE kernel."
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Does warning here need to be fixed? Since we are using flashinfer trtllm runner

File: python/sglang/srt/server_args.py [hunk 4]
```diff
@@ -936,7 +935,9 @@ def _handle_model_specific_adjustments(self):
             )
 
             if is_blackwell_supported() and is_mxfp4_quant_format:
-                self.moe_runner_backend = "flashinfer_mxfp4"
+                if self.moe_runner_backend == "auto":
+                    self.moe_runner_backend = "flashinfer_trtllm"
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Is it OK to totally replace with `flashinfer_trtllm` ?
Since there are models also using `flashinfer-trtllm` backend (dpsk nvfp4), and they might require different weight loading/preparation

File: test/srt/test_gpt_oss_4gpu.py [hunk 5]
```diff
@@ -31,6 +33,60 @@ def test_mxfp4_120b(self):
             ],
         )
 
+    @unittest.skipIf(get_device_sm() < 100, "Test requires CUDA SM 100 or higher")
+    def test_mxfp4_trtllm_moe_120b(self):
```
Comment (Fridge003) [inline test/srt/test_gpt_oss_4gpu.py]: Can we open a new test file `test_gpt_oss_backend_4gpu.py` and put these tests there?
Also can we skip them in CI, since currently we only have one B200 for CI. This will add a lot of burden. In the future we can offload them to nightly tests.

File: python/sglang/srt/server_args.py [hunk 6]
```diff
@@ -1343,13 +1344,20 @@ def _handle_moe_kernel_config(self):
...
+                or self.quantization == "mxfp4"
                 or self.quantization == "fp8"
-            ), "modelopt_fp4, modelopt_fp8 or fp8 quantization is required for Flashinfer TRTLLM MoE"
+            ), "modelopt_fp4, modelopt_fp8, mxfp4 or fp8 quantization is required for Flashinfer TRTLLM MoE"
             self.disable_shared_experts_fusion = True
             logger.warning(
                 "FlashInfer TRTLLM MoE is enabled. --disable-shared-experts-fusion is automatically set."
             )
 
+        if self.moe_runner_backend == "flashinfer_mxfp4":
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Can move this to `_handle_deprecated_args`

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 7]
```diff
@@ -467,6 +473,20 @@ def _map_global_expert_id_to_local_expert_id(self, expert_id: int) -> int:
...
+        self,
+        param: torch.nn.Parameter,
+        loaded_weight: torch.Tensor,
+        weight_name: str,
+    ) -> None:
+        dim1 = loaded_weight.shape[1]
+        if "bias" in weight_name:
+            param.data[:, :dim1].copy_(loaded_weight)
+        else:
+            packing_factor = param.data.element_size() // loaded_weight.element_size()
```
Comment (Fridge003) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py:512]: When is `packing_factor` not equal to 1?
The original code didn't consider this packing_factor, so this makes me wonder

File: python/sglang/srt/layers/moe/topk.py [hunk 8]
```diff
@@ -256,10 +256,7 @@ def forward_cuda(
             output_format = self.topk_config.output_format
         elif get_moe_runner_backend().is_triton_kernels():
             output_format = TopKOutputFormat.TRITON_KERNEL
-        elif (
-            should_use_flashinfer_trtllm_moe()
-            or get_moe_runner_backend().is_flashinfer_mxfp4()
-        ):
+        elif should_use_flashinfer_trtllm_moe():
```
Comment (Fridge003) [inline python/sglang/srt/layers/moe/topk.py]: Will this change affect other model using flashinfer_trtllm?

File: python/sglang/srt/layers/quantization/mxfp4.py [hunk 9]
```diff
@@ -79,6 +81,11 @@
             e8m0_shuffle
         ) = err
 
+try:
```
Comment (Fridge003) [inline python/sglang/srt/layers/quantization/mxfp4.py]: Can we lazily import this function (just before calling)

File: python/sglang/srt/server_args.py (hunk 3 above)
Comment (nvcastet) [reply python/sglang/srt/server_args.py]: Yes I can update here, it is using the same runner by default but the warning is not precise.

File: python/sglang/srt/server_args.py (hunk 4 above)
Comment (nvcastet) [reply python/sglang/srt/server_args.py]: Yes it is.
So both `flashinfer_trtllm` and `flashinfer_cutlass` backends support different kernels for different quantizations (fp4, fp8, mxfp4, unquant).
So kernels are selected via the tuple (backend, quantization), so no reason to bake the quantization in the backend name.

File: test/srt/test_gpt_oss_4gpu.py (hunk 5 above)
Comment (nvcastet) [reply test/srt/test_gpt_oss_4gpu.py]: Done

File: python/sglang/srt/server_args.py (hunk 6 above)
Comment (nvcastet) [reply python/sglang/srt/server_args.py]: Done

File: python/sglang/srt/layers/moe/topk.py (hunk 8 above)
Comment (nvcastet) [reply python/sglang/srt/layers/moe/topk.py]: No it is just cleaning up to keep quantization out of the backend name.

File: python/sglang/srt/layers/quantization/mxfp4.py (hunk 9 above)
Comment (nvcastet) [reply python/sglang/srt/layers/quantization/mxfp4.py]: done

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 7 above)
Comment (nvcastet) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py:512]:
```
        self.weight_dtype = (
            torch.int64 if self.enable_flashinfer_cutlass_moe else torch.uint8
        )
```

For cutlass weight dtype is 64bit. For trtllm-gen, weight dtype matches the checkpoint dtype (8bits) so factor is always 1.

File: test/srt/run_suite.py [hunk 10]
```diff
@@ -304,6 +304,7 @@ class TestFile:
         TestFile("test_fp8_kvcache.py"),
         TestFile("test_full_deepseek_v3.py"),
         TestFile("test_get_weights_by_name.py"),
+        TestFile("test_gpt_oss_backend_4gpu.py"),
```
Comment (merrymercy) [inline test/srt/run_suite.py]: move this to gb200 or b200 ci?

Comment (nvcastet) [reply test/srt/run_suite.py]: https://github.com/sgl-project/sglang/pull/9667#discussion_r2484174572

^^ I just moved it out with @Fridge003 recommendation

Comment (Fridge003) [reply test/srt/run_suite.py]: @nvcastet Now we can move it to nightly b200 test
https://github.com/sgl-project/sglang/blob/main/test/srt/run_suite.py#L218

Comment (nvcastet) [reply test/srt/run_suite.py]: Done

**Comment by Fridge003** [comment]
@harvenstar Thanks, just manually applied your change

**Review by Fridge003 — APPROVED** [review]

**Comment by Fridge003** [comment]
Noticed some throughput regression... @harvenstar will check this
